core/trino-main/src/main/java/io/trino/operator/join/JoinHashSupplier.java

```diff
@@ -79,9 +82,29 @@ public JoinHashSupplier(
             positionLinksFactoryBuilder = ArrayPositionLinks.builder(addresses.size());
         }
 
-        this.pages = channelsToPages(channels);
+        long additionalPagesRetainedSizeInBytes = 0;
+        // Currently PageIndex is retained by the HashBuilderOperator after JoinHash is created.
+        // JoinHash has to account for all data structures retained by PageIndex as after JoinHash is created memory retained by PageIndex is no longer accounted for.
```

Comment:
Also can we just see how much much memory `PageIndex` consumed instead recomputing piece by piece here?

Reply:
`PageIndex` stores data as a `List<List<Block>>`. However `JoinHash` nees the same data in a form of `List<Page>`, so we have to account the overhead of `Page` wrappers, but avoid accounting the size of blocks twice. Ideally it would be great to refactor and only use data in one form.